python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,824 @@
+def explore(data, bounds, current, base, old, config, mutations, resample_probability):
+        
+    data['Trial'] = data['Trial'].astype(str)
+    data.to_csv('old.csv')
+
+    df = data.sort_values(by='T').reset_index(drop=True)
+
+    # group by trial ID and hyperparams. Compute change in timesteps and reward. 
+    df['y'] = df.groupby(['Trial'] + [key for key in bounds.keys()])['Reward'].diff()
+    df['t_change'] = df.groupby(['Trial'] + [key for key in bounds.keys()])['T'].diff() 
+
+    # delete entries without positive change in t. 
+    df = df[df['t_change'] > 0].reset_index(drop=True)
+    df['R_before'] = df.Reward - df.y
+
+    # normalize the reward change by the update size. For example if trials took diff lengths of time.
+    df['y'] = df.y / df.t_change
+    df = df[~df.y.isna()].reset_index(drop=True)
+    df = df.sort_values(by='T').reset_index(drop=True)
+
+    # only use the last 1k datapoints, so the GP is not too slow
+    df = df.iloc[-1000:, :].reset_index(drop=True) 
+    
+    # we need this to know the T and Reward for the weights 
+    dfnewpoint = df[df['Trial']==str(base)]
+    
+    if not dfnewpoint.empty:
+                
+        # now specify the dataset for the GP
+        y = np.array(df.y.values)
+        t_r = df[['T', 'R_before']] # we use the T and starting reward as features
+        h = df[[key for key in bounds.keys()]]
+        X = pd.concat([t_r, h], axis=1).values 
+        newpoint = df[df['Trial']==str(base)].iloc[-1, :][['T', 'R_before']].values
+        new = select_config(X,y, current, newpoint, bounds, num_f = len(t_r.columns), num=len(h.columns))
+
+        new_config = config.copy()
+        values = []
+        names = []
+        for i, col in enumerate(h.columns):
+            if type(config[col]) is int:
+                new_config[col] = int(new[i])
+                values.append(int(new[i]))
+            else:
+                new_config[col] = new[i]
+                values.append(new[i])
+                
+        new_T = df[df['Trial']==str(base)].iloc[-1, :]['T']
+        new_Reward = df[df['Trial']==str(base)].iloc[-1, :].Reward
+              
+        lst = [[old] + [new_T]  + values + [new_Reward]]
+        cols = ['Trial', 'T'] + [key for key in bounds.keys()] + ['Reward']
+        new_entry = pd.DataFrame(lst, columns = cols) 
+
+        # create an entry for the new config, with the reward from the copied agent.
+        data = pd.concat([data, new_entry]).reset_index(drop=True)
+
+    else:
+        new_config = config.copy()
+
+    return new_config, data
+
+
+def make_experiment_tag(orig_tag, config, mutations):
+    """Appends perturbed params to the trial name to show in the console."""
+
+    resolved_vars = {}
+    for k in mutations.keys():
+        resolved_vars[("config", k)] = config[k]
+    return "{}@perturbed[{}]".format(orig_tag, format_vars(resolved_vars))
+
+
+class PB2(FIFOScheduler):
```

Comment:
@jparkerholder Does it make sense to subclass ``PopulationBasedTraining`` here to avoid having to repeat a lot of the code?

Reply:
that makes a lot of sense!

Reply:
I did my best to make this change - hopefully it is what you were thinking :)

Reply:
I am not at all an expert on hierarchy, but the current version with super(PopulationBasedTraining, self).init(...) leads to errors in my tests. I get the following:

```
line 285, in __init__
    custom_explore_fn, log_config, require_attrs, synch)
TypeError: object.__init__() takes no arguments
```

Any ideas? I can share the script I am using for testing if that helps. It worked prior to this change.

Reply:
My fault! This should be working now.